### ISSUE TITLE
Extending the code to new video models

### ISSUE THREAD

**Issue by thechargedneutron** [XueZeyue/DanceGRPO#69]
Hi,

Thanks for the amazing work. I want to extend this GRPO training to newer video models like Wan, Cog, LTX etc. I have some questions:

Firstly, I see that you have pipeline code for `hunyuan_hf` in [this](https://github.com/XueZeyue/DanceGRPO/tree/main/fastvideo/models/hunyuan_hf) folder. It looks very similar to the pipeline code in diffusers. Have you simply copied it, or did you make any changes to it? Doing a simple diff gives more differences, but I suspect it's only due to update in diffusers code.

Next, do you know if the code changes have anything specific with hunyuan that I should be careful about when plugging a different model like Wan?

Thanks in advance

**Comment by thechargedneutron** [comment]
Thanks for the prompt reply.

> the pipeline code is copied from FastVideo.

I don't see similar code in the FastVideo file. The closest match is get is this file: https://github.com/hao-ai-lab/FastVideo/blob/main/fastvideo/pipelines/basic/hunyuan/hunyuan_pipeline.py but this looks very different, whereas the diffusers implementation available [here](https://github.com/huggingface/diffusers/blob/c934720629837257b15fd84d27e8eddaa52b76e6/src/diffusers/pipelines/hunyuan_video/pipeline_hunyuan_video.py) is closer. Can you point me to the appropriate place in FastVideo from where the pipeline code given [here](https://github.com/XueZeyue/DanceGRPO/blob/main/fastvideo/models/hunyuan_hf/pipeline_hunyuan.py) is obtained from?

**Comment by XueZeyue** [comment]
Hi, I use a relatively older version to develop DanceGRPO (https://github.com/hao-ai-lab/FastVideo/tree/218449c54d400d91a6346d9602d717fe725b0d3f/fastvideo/models/hunyuan_hf)
It seems that FastVideo adds some sp operations to it, such as:
```python
from fastvideo.utils.communications import all_gather, all_to_all_4D
from fastvideo.utils.parallel_states import (get_sequence_parallel_state,
                                             nccl_info)
```

**Comment by thechargedneutron** [comment]
Thanks for pointing this out, very helpful!

> BTW, wan-2.1-1.3b doesn't need FSDP and SP; you just need to apply gradient checkpointing, and it's very convenient to develop.

Actually, I want to develop the 14B variant, and also some other video generation models. Hence, putting in the effort to enable FSDP and SP there as well.

**Comment by thechargedneutron** [comment]
@XueZeyue  Hi, I realized Wan uses a different scheduler, i.e., UniPCMultistepScheduler. Do you think we can use the same scheduler as used in HunyuanVideo to sample videos and train a GRPO model? I did minor modifications to enable Wan2.1-1.3B, but I am getting outputs as noted in #59

**Comment by thechargedneutron** [comment]
Thanks, this is really helpful. I did all the other required changes, along with this, and now I see much better output. However, the current code does not implement classifier-free guidance [here](https://github.com/huggingface/diffusers/blob/c934720629837257b15fd84d27e8eddaa52b76e6/src/diffusers/pipelines/wan/pipeline_wan.py#L541-L549). I checked Wan output, and they look much better with classifier-free guidance (CFG). Do you think this GRPO setting can be extended to CFG?

**Comment by XueZeyue** [comment]
Hi, for training with cfg, you can refer to the SkyReels-I2V training settings.
You should set gradient_accumulation_steps = train_batch_size*num_generations, which means you should update the gradients only once in each iteration.

**Comment by thechargedneutron** [comment]
@XueZeyue Hi, thanks for the help with Wan. Is the implementation of `mochi` also done? I don't see a `train_grpo_mochi.py` code, but I see `mochi_hf` in the fastvideo folder. Is that fully implemented?
